go/vt/vtgate/evalengine/fn_string.go

```diff
@@ -120,6 +127,122 @@ var _ IR = (*builtinLeftRight)(nil)
 var _ IR = (*builtinPad)(nil)
 var _ IR = (*builtinTrim)(nil)
 
+func insert(str, newstr *evalBytes, pos, l int) []byte {
+	pos--
+
+	cs := colldata.Lookup(str.col.Collation).Charset()
+	strLen := charset.Length(cs, str.bytes)
+
+	if pos < 0 || strLen <= pos {
+		return str.bytes
+	}
+	if l < 0 {
+		l = strLen
+	}
+
+	front := charset.Slice(cs, str.bytes, 0, pos)
+	var back []byte
+	if pos <= math.MaxInt-l && pos+l < strLen {
+		back = charset.Slice(cs, str.bytes, pos+l, strLen)
+	}
+
+	res := make([]byte, len(front)+len(newstr.bytes)+len(back))
+
+	copy(res[:len(front)], front)
+	copy(res[len(front):], newstr.bytes)
+	copy(res[len(front)+len(newstr.bytes):], back)
```

Comment:
@vmg Do you think it's worth extending `charset.Slice` to allow passing in an existing output `[]byte` or is that a bit premature just for this function?

Reply:
I think it would make for a more awkward API, don't you think? Right now `Slice` doesn't allocate (it returns a sub-slice of the input), so if we add an `out` parameter, we'd have to make it always allocate _or_ have two code paths in the function for when `out` is `nil`. 

Furthermore, if the `Slice` API had an `out` parameter, it would allocate more than once, since we wouldn't know the required capacity of the `out` buffer, so it'd have to grow it at least twice.

With the current non-allocating `Slice`, we just call the function twice and then we get to compose the final result outside of the function with a single allocation, since we know the final size. I quite dig the version of the code as it is written.

Reply:
Ah yeah, good point. It's already non-allocating so it doesn't really help in any way.